graphjet-core/src/main/java/com/twitter/graphjet/algorithms/counting/TopSecondDegreeByCount.java

```diff
@@ -14,45 +14,42 @@
  * limitations under the License.
  */
 
-
 package com.twitter.graphjet.algorithms.counting;
 
-import java.util.ArrayList;
-import java.util.List;
-import java.util.Random;
-
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
-
 import com.twitter.graphjet.algorithms.NodeInfo;
 import com.twitter.graphjet.algorithms.RecommendationAlgorithm;
-import com.twitter.graphjet.algorithms.RecommendationInfo;
 import com.twitter.graphjet.algorithms.RecommendationStats;
-import com.twitter.graphjet.algorithms.RecommendationType;
 import com.twitter.graphjet.bipartite.NodeMetadataLeftIndexedMultiSegmentBipartiteGraph;
 import com.twitter.graphjet.bipartite.NodeMetadataMultiSegmentIterator;
-import com.twitter.graphjet.hashing.IntArrayIterator;
 import com.twitter.graphjet.stats.Counter;
 import com.twitter.graphjet.stats.StatsReceiver;
+import it.unimi.dsi.fastutil.longs.*;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
-import it.unimi.dsi.fastutil.longs.Long2ByteArrayMap;
-import it.unimi.dsi.fastutil.longs.Long2ByteMap;
-import it.unimi.dsi.fastutil.longs.Long2DoubleMap;
-import it.unimi.dsi.fastutil.longs.Long2ObjectMap;
-import it.unimi.dsi.fastutil.longs.Long2ObjectOpenHashMap;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
 
-public class TopSecondDegreeByCount implements
-    RecommendationAlgorithm<TopSecondDegreeByCountRequest, TopSecondDegreeByCountResponse> {
-  private static final Logger LOG = LoggerFactory.getLogger("graph");
-  private static final int MAX_EDGES_PER_NODE = 500;
+/**
+ * Generates recommended RHS nodes by calculating aggregated weights.
+ * Weights are accumulated by weights of LHS nodes whose edges are incident to RHS nodes
+ */
+public abstract class TopSecondDegreeByCount<Request extends TopSecondDegreeByCountRequest,
+  Response extends TopSecondDegreeByCountResponse>
+  implements RecommendationAlgorithm<Request, Response> {
+
+  protected static final Logger LOG = LoggerFactory.getLogger("graph");
+  protected static final int MAX_EDGES_PER_NODE = 500;
 
+  // Static variables for better memory reuse. Avoids re-allocation on every request
   private final NodeMetadataLeftIndexedMultiSegmentBipartiteGraph leftIndexedBipartiteGraph;
-  private final Long2ObjectMap<NodeInfo> visitedRightNodes;
-  private final List<NodeInfo> nodeInfosAfterFiltering;
   private final Long2ByteMap seenEdgesPerNode;
-  private final RecommendationStats topSecondDegreeByCountStats;
-  private final StatsReceiver statsReceiver;
-  private final Counter numRequestsCounter;
+  protected final Long2ObjectMap<NodeInfo> visitedRightNodes;
+  protected final List<NodeInfo> nodeInfosAfterFiltering;
+  protected final RecommendationStats topSecondDegreeByCountStats;
+  protected final StatsReceiver statsReceiver;
+  protected final Counter numRequestsCounter;
 
   /**
    * This initializes all the state needed to run TopSecondDegreeByCount. Note that the object can
```

Comment:
even if we initialize all the states, we can not run TopSecondDegreeByCount because it is an abstract class, right?